fl4health/clients/moon_client.py

```diff
@@ -0,0 +1,107 @@
+import copy
+from pathlib import Path
+from typing import Dict, List, Optional, Sequence
+
+import torch
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
+from fl4health.clients.basic_client import BasicClient
+from fl4health.model_bases.moon_base import MoonModel
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
+
+
+class MoonClient(BasicClient):
+    """
+    This client implements the MOON algorithm from Model-Contrastive Federated Learning. The key idea of MOON
+    is to utilize the similarity between model representations to correct the local training of individual parties,
+    i.e., conducting contrastive learning in model-level.
+    """
+
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.initial_tensors: List[torch.Tensor]
+        self.contrastive_weight: float = 10
```

Comment:
We will probably want the ability to configure hyperparameters (temperature, contrastive weight and len_old_models_buffer) to take advantage of your otherwise general code. We can have the defaults that you specify but check if the config sent by the server contains these hyperparams, and if they do set them accordingly before the FL round begins.